### ISSUE TITLE
numpy >= 1.16 with pypy and ctypes errors

### ISSUE THREAD

**Issue by texasaggie97** [numpy/numpy#13807]
### Reproducing code example:
You can see an example of this failure on [Travis CI](https://travis-ci.org/ni/nimi-python/jobs/547906996#L1381)
```python
import numpy as np
return np.ctypeslib.as_ctypes(value)
```
where value is a numpy array with dtype='float64'

This works when restricting nimpy to '<1.16' on pypy and the latest numpy works with this on CPython.

While the examples I have put in this issue refer to pypy 2.7.13-7.1.1, this also applies to pypy 3.6.1-7.1.1

### Error message:

```
bin/nifake/nifake/session.py:72: in aux
    return f(*xs, **kws)
bin/nifake/nifake/session.py:1613: in write_waveform_numpy
    waveform_ctype = get_ctypes_pointer_for_buffer(value=waveform)  # case B510
bin/nifake/nifake/session.py:31: in get_ctypes_pointer_for_buffer
    return numpy.ctypeslib.as_ctypes(value)
.tox/64/pypy-test/site-packages/numpy/ctypeslib.py:539: in as_ctypes
    result = as_ctypes_type(dtype).from_address(addr)
.tox/64/pypy-test/site-packages/numpy/ctypeslib.py:502: in as_ctypes_type
    return _ctype_from_dtype(_dtype(dtype))
.tox/64/pypy-test/site-packages/numpy/ctypeslib.py:459: in _ctype_from_dtype
    return _ctype_from_dtype_subarray(dtype)
.tox/64/pypy-test/site-packages/numpy/ctypeslib.py:397: in _ctype_from_dtype_subarray
    ctype = _ctype_from_dtype(element_dtype)
.tox/64/pypy-test/site-packages/numpy/ctypeslib.py:461: in _ctype_from_dtype
    return _ctype_from_dtype_scalar(dtype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dtype = dtype('float64')
    def _ctype_from_dtype_scalar(dtype):
        # swapping twice ensure that `=` is promoted to <, >, or |
        dtype_with_endian = dtype.newbyteorder('S').newbyteorder('S')
        dtype_native = dtype.newbyteorder('=')
        try:
            ctype = _scalar_type_map[dtype_native]
        except KeyError:
            raise NotImplementedError(
>               "Converting {!r} to a ctypes type".format(dtype)
            )
E           NotImplementedError: Converting dtype('float64') to a ctypes type```
```

### Numpy/Python version information:

```
OS:
    Name:      Linux
    Version:   #34~16.04.1-Ubuntu SMP Wed May 8 13:26:26 UTC 2019
    Bits:      64
Python:
    Version:   2.7.13 (8cdda8b8cdb8, Apr 14 2019, 14:06:44)
[PyPy 7.1.1 with GCC 6.2.0 20160901]
    Bits:      64
    Is_Venv:   True
    Installed Packages:
        nifake==1.1.3.dev0
        readline==6.2.4.1
        greenlet==0.4.13
        cffi==1.12.3
        wsgiref==0.1.2
        zipp==0.5.1
        wheel==0.33.1
        wcwidth==0.1.7
        six==1.12.0
        singledispatch==3.4.0.3
        setuptools==40.8.0
        scandir==1.10.0
        pytest==4.6.3
        pyparsing==2.4.0
        py==1.8.0
        pluggy==0.12.0
        pip==19.1.1
        pathlib2==2.3.3
        packaging==19.0
        numpy==1.16.4
        more-itertools==5.0.0
        mock==3.0.5
        markupsafe==1.1.1
        mako==1.0.12
        importlib-metadata==0.18
        funcsigs==1.0.2
        enum34==1.1.6
        coverage==4.5.3
        contextlib2==0.5.5
        configparser==3.7.4
        attrs==19.1.0
        atomicwrites==1.3.0
```

**Comment by mattip** [comment]
ctypes support improved after the release of PyPy 7.1.1. Numpy tests with a [nightly build](https://github.com/numpy/numpy/blob/master/tools/pypy-test.sh#L32).

**Comment by eric-wieser** [comment]
Can you print out dtype.char here?

**Comment by eric-wieser** [comment]
Thanks, I wanted to confirm the failure was not long double

**Comment by eric-wieser** [comment]
@matti: this looks like a hashing/ dict lookup problem, not a ctypes problem

**Comment by mattip** [comment]
Can reproduce on latest PyPY with 1.16.4, but not with HEAD. The problem is that `np.dtype(ctypes.c_int)` produces `dtype('O')` on 1.16.4, `dtype('int32') on HEAD. Something was fixed and not backported, but I don't remember what.

**Comment by eric-wieser** [comment]
`np.ctypeslib.as_ctypes_type(ctypes.c_int)` is a simpler test. Can you run that on both head and 1.16.4?

**Comment by mattip** [comment]
If we wish to fix this we should backport e67f8c7f to 1.16. It was part of PR #12594 to add PyPY3 to our CI, but 1.16.4 is the LTS python2 branch.

**Comment by mattip** [comment]
Closing. The fix to this will be part of v1.16.5, the next LTS release for Python2. Please open a new issue if there are still issues with PyPy and the 1.16.x HEAD
